Limits steps in onboarding based on previously completed state

dsawardekar opened the pull request:
### Changes proposed in this Pull Request:

Closes #2493 

This PR hides the confirm registration step if the user has sufficient data filled in already. If the user was on the confirm registration step previously, they are sent to the next step.


### Screenshots:

With incomplete store requirements: (Confirm store requirements is present)
![image](https://github.com/user-attachments/assets/bcdca583-0252-4c42-939b-0400a2305072)


With complete store requirements: (Confirm store requirements is absent)
![image](https://github.com/user-attachments/assets/85416575-29bb-40e9-85c8-fd1abedccd80)


### Detailed test instructions:


1. Start the onboarding process with incomplete store data.
2. You should see the Step no. 3 Confirm Store Requirements
3. Reset your env, and delete some store data (eg:- Delete City under WooCommerce Store Address)
4. Restart the onboarding process
5. You should not see the Step no. 3


### Additional details:

Update: Hides the Confirm Store Requirements step when feasible

### Changelog entry

Update: Hides the Confirm Store Requirements step when feasible

dsawardekar committed:
Limits steps in onboarding based on previously completed state

dsawardekar committed:
Fixes linter warnings

dsawardekar commented:
@joemcgill This is working but I needed to move the implementation to the parent container component. Having the logic inside the SavedSetupStepper was causing a broken UX like below when the user already had completed store requirements.

![Screenshot from 2024-08-21 16-31-16](https://github.com/user-attachments/assets/6b58d8a8-1451-440f-ba85-e2ee16f16956)

Can you review and let me know if this looks good to proceed on the e2e tests? 

Thanks!

joemcgill commented on js/src/setup-mc/setup-stepper/index.js:
```diff
@@ -9,29 +9,60 @@ import { getHistory, getNewPath } from '@woocommerce/navigation';
+	const { status } = mcSetup;
+	let { step } = mcSetup;
+
+	// If the user has already completed the store requirements, but is currently still on the
+	// store requirements step, we should skip the store requirements step and go to the paid ads step.
+	// else they will get stuck on a non-existent step #3
+	if ( step === 'store_requirements' && hasConfirmedStoreRequirements ) {
+		step = 'paid_ads';
+	}
```
Smart. Rather than mutating the `step` variable here, it may be better to do something like this:

```js
const { status, step } = mcSetup;

const currentStep = ( step === 'store_requirements' && hasConfirmedStoreRequirements ) ?  'paid_ads' : step;
```

joemcgill reviewed (commented):
@dsawardekar I think this approach makes sense. Particularly since it's the `SetupStepper` that manages the loading state prior to rendering the `SavedSetupStepper`.

Once you've updated the E2E tests, feel free to send this back for review.

dsawardekar committed:
Updates step conditional per feedback

dsawardekar committed:
First pass at the hide store requirements e2e tests

dsawardekar committed:
Fixes linter warning

dsawardekar committed:
Switches to complete step mock

dsawardekar commented:
@joemcgill I've added the e2e tests for this, but I need assistance with mocking the email and mc account address. When I try to set mockMCSetup to use 'complete', it redirects to the Home page, and skips the onboarding flow completely.

![image](https://github.com/user-attachments/assets/2c8b52df-2a0b-49b6-bd40-f6e64e2c15ce)


Test File:
https://github.com/woocommerce/google-listings-and-ads/blob/31a3ba5017434a3d4f6995f89839dd64546eb888/tests/e2e/specs/setup-mc/step-3-hide-store-requirements.test.js

Can you please take a look? Thanks!

joemcgill commented on js/src/setup-mc/setup-stepper/index.js:
```diff
@@ -43,14 +43,15 @@ const SetupStepper = () => {
 
 	const { status } = mcSetup;
-	let { step } = mcSetup;
+	const { step } = mcSetup;
```
These can be combined:

```js
const { status, step } = mcSetup;
```

joemcgill commented on tests/e2e/specs/setup-mc/step-3-hide-store-requirements.test.js:
Rather than creating a whole new spec here, I think we should test for these scenarios in either the existing `tests/e2e/specs/setup-mc/step-3-confirm-store-requirements.test.js` spec, or test for this scenario in `tests/e2e/specs/setup-mc/step-2-product-listings.test.js` whenever the continue button is clicked with the address info already correctly mocked. I think that should look something like this:

```js
productListingsPage.mockContactInformation( {
	phone_number: '+15555555',
	phone_verification_status: 'verified',
	mc_address: {
		street_address: '556 Woo St.',
		locality: 'City',
		region: 'California',
		postal_code: '90210',
		country: 'US',
	},
	wc_address: {
		street_address: '556 Woo St.',
		locality: 'City',
		region: 'California',
		postal_code: '90210',
		country: 'US',
	},
	is_mc_address_different: false,
	wc_address_errors: [],
} );
```

One tricky part here is that the step is only removed when the stepper is first loaded, so we'll need a beforeAll step that reloads the page after the state is correctly mocked.

I'm going to take a look tomorrow to see if I can get the state for this set up properly, since it's a bit tricky.

joemcgill reviewed (commented):
Thanks @dsawardekar. I've added a bit of initial feedback. I'm going to see if I can get these e2e tests working, but in the mean time can you address the other feedback?

joemcgill reviewed (commented):
@dsawardekar while testing this again, I've realized that part of the problem that has made the E2E difficult to debug is that these updates won't work when an MC account is not connected when the onboarding flow is loaded, which is a pretty significant problem.

If you test this with all accounts disconnected, you'll see an error message pop up when the following hooks are called:

https://github.com/woocommerce/google-listings-and-ads/blob/abc86c3a2ce2a6813c6da38d7707b6b70ff1934f/js/src/setup-mc/setup-stepper/index.js#L18-L19

Sorry I didn't catch this when you initially suggested moving these hooks here. I assume I was testing by just refreshing the page after already completing step 1 of the onboarding. I'm curious if we're able to initially show step 3 and then hide it once someone connects their MC account in step 1. If not, then we'll likely need to rethink this issue.

dsawardekar commented:
@joemcgill That might work, but the UX would be less than ideal. Maybe we need to consider folding the things that in that extra dissapearing step into sections inside Step 2 / 3. Then if the user has sufficient data, we could hide that section.

joemcgill committed:
Check store requirements in saved setup stepper.

This updates both the `useStoreAddress` and `useGoogleMCPhoneNumber` hooks to check for a connected MC account before fetching contact information so the saved setup stepper can determine if the user has the necessary information to proceed prior to connecting an MC account.

eason9487 commented:
Closing this as an alternative has been implemented and merged. See https://github.com/woocommerce/google-listings-and-ads/issues/2493#issuecomment-2391705651.
